src/pseudo/PspUpf.jl

```diff
@@ -0,0 +1,234 @@
+    vloc_interp                       # Local potential interpolator
+    
+    projs::Vector{Vector{Vector{T}}}  # Kleinman-Bylander β projectors: projs[l+1][i]
+    projs_interp                      # Projector interpolator
+    h::Vector{Matrix{T}}              # Projector coupling coefficients per AM channel: h[l+1][i1,i2]
+    
+    pswfcs::Vector{Vector{T}}         # Pseudo-atomic wavefunctions
+    pswfc_ang_moms::Vector{Int}       # Angular momenta of the pseudo-atomic wavefunctions
+    pswfc_occs::Vector{T}             # Occupations of the pseudo-atomic wavefunctions
+
+    rvlocpzdr::Vector{T}                # (rgrid * vloc + Z) * drgrid
+    rprojsdr::Vector{Vector{Vector{T}}} # rgrid * projs[l+1][i] * drgrid
+    
+    identifier::String                # String identifying the PSP
+end
+
+"""
+    PspUpf(Zion::Number, lmax::Number, rgrid::Vector, vloc::Vector,
+           projs::Vector{Vector{Vector}}, h::Vector{Matrix}; identifier="")
+
+Construct a Unified Pseudopotential Format pseudopotential. Currently only
+norm-conserving potentials are supported.
+"""
+function PspUpf(Zion, lmax, rgrid::Vector{T}, drgrid, vloc, vloc_interp, projs, projs_interp, h,
+                pswfcs, pswfc_ang_moms, pswfc_occs, rvlocpzdr, rprojsdr; identifier="") where {T}
+    length(projs) == length(h) || error("Length of projs and h do not agree.")
+    length(rprojsdr) == length(h) || error("Length of rgrid * projs * dr and h do not agree.")
+    length(pswfcs) == length(pswfc_ang_moms) || error("Length of pseudo wfcs and pseudo wfc" *
+                                                      "angular momenta do not agree.")
+    length(pswfcs) == length(pswfc_occs) || error("Length of pseudo wfcs and pseudo wfc" *
+                                                      "occupations do not agree.")
+
+    return PspUpf{T}(Zion, lmax, rgrid, drgrid, vloc, vloc_interp, projs, projs_interp, h,
+                     pswfcs, pswfc_ang_moms, pswfc_occs, rvlocpzdr, rprojsdr, identifier)
+end
+
+function parse_upf_file(path; identifier=path)
+    pseudo = load_upf(path)
+
+    unsupported = []
+    pseudo["header"]["core_correction"] && push!(unsupported, "nonlinear core correction")
+    pseudo["header"]["has_so"] && push!(unsupported, "spin-orbit coupling")
+    pseudo["header"]["pseudo_type"] == "SL" && push!(unsupported, "semilocal potential")
+    pseudo["header"]["pseudo_type"] == "US" && push!(unsupported, "ultrasoft")
+    pseudo["header"]["pseudo_type"] == "PAW" && push!(unsupported, "plane-augmented wave")
+    pseudo["header"]["has_gipaw"] && push!(unsupported, "gipaw reconstruction data")
+    pseudo["header"]["pseudo_type"] == "1/r" && push!(unsupported, "Coulomb")
+    length(unsupported) > 0 && error("Pseudopotential contains the following unsupported "
+                                        * "features/quantities: $unsupported")
+
+    # Pseudo-ion charge
+    Zion = Int(pseudo["header"]["z_valence"])
```

Comment:
Having the comments before the extraction sort of destroys the flow. Maybe have it in the same line after?